### ISSUE TITLE
Wrong (according to the article) hidden state computation

### ISSUE THREAD

**Issue by andreyzharkov** [datalogue/keras-attention#27]
Hi!
Thank you for the code implementing attention, it was really helpful.

You are referencing the article https://arxiv.org/pdf/1409.0473.pdf
But it seems that your implementation differs from that in the article when you are computing hidden state (proposed).
The original formulas in the article: 
![image](https://user-images.githubusercontent.com/14358106/37031933-d496103c-2151-11e8-80f0-8a1a97c68e6b.png)
Pay attention at the weights before y_{i-1}, W - is the weights matrix but **E - is embedding matrix**

Now let's look what is impremented in AttentionDecoder:
```
        # now calculate the "z" gate
        zt = activations.sigmoid(
            K.dot(ytm, self.W_z)
            + K.dot(stm, self.U_z)
            + K.dot(context, self.C_z)
            + self.b_z)

        # calculate the proposal hidden state:
        s_tp = activations.tanh(
            K.dot(ytm, self.W_p)
            + K.dot((rt * stm), self.U_p)
            + K.dot(context, self.C_p)
            + self.b_p)

        # new hidden state:
        st = (1-zt)*stm + zt * s_tp

        yt = activations.softmax(
            K.dot(ytm, self.W_o)
            + K.dot(stm, self.U_o)
            + K.dot(context, self.C_o)
            + self.b_o)
```
So instead of previous output word `ytm` is the softmax probability of the labels distribution of the previous timestep. And that probabilities is passed into further computation instead of embedding vector.

This is a bug, at least it can easily lead to overfitting - the sofrmax probabilities give much more information than the embedding of the previous word. Moreover the size of one of the weights is (output_dim x output_dim). I was suffered from it even with the small character vocabulary (hieroglyphs) of about 3000, and can't imagine that this code may be used to provide translation with vocabe sizes of hungreds of thousands.

Hope this will be fixed soon

**Comment by zafarali** [comment]
You're infact correct that the paper uses the true labels. This is known as [teacher forcing](http://www.deeplearningbook.org/contents/rnn.html) and is meant to make training more stable. It is, however, optional and I wanted to try the simplest thing first and it seemed to work for the problem I was considering it.

You are correct that I should be using the embedding matrix there, however, with the current architectural setup it is unclear if this is possible using vanilla keras.

**Comment by andreyzharkov** [comment]
@zafarali, thank you for reply. Now I understand your motivation better. But maybe you should mention the differences with original paper in readme or in `AttentionDecoder` docstring? Because now it is a bit misleading.

I agree that it will be hard to implement with keras only. But you are using tf in cursom_recurrents.py (although it seems to be the only place and in fact unused). So tf is one of the dependencies now and you may use it to implement embeddings. But more likely it will make the code less readable, and, again, tf is in fact unused, so I just wanted to mention the opportunity.

Also, it is likely that I was incorrect writing that softmax probabilities may be used to extract more information. Of course, they give a bit more information than one-hot, but training the model with cross-entropy will stimulate probabilities to approach one-hot as much as possible. So maybe it is not a major problem.

So, **the only problem that remains** (in my view) is that the implemented model is hard to be used with large vocabulary sizes. In my case (3800) the proposed attention model gives about 20M trainable weights, while similar model with CTC instead of attention has only 3M. So the model is harder to train and it will be inefficient as inference. I'm not insisting on changing something to deal with that problem (howewer, it will be great), but maybe you can add this information into readme or somewhere else to help people understand why they have so many weights (actually it was the reason why I started investigating the implementation).

**Comment by zafarali** [comment]
Regarding the first point, yes I agree it is misleading. Could you send a pull request with an updated docstring and README? In the readme you could add a section saying _”Differences between original paper and implementation”_ and list the teacher forcing issue as well as the max-out network?

Regarding the second problem, I think the reason for the model increasing in parameters is because of the recurrent network and not the attention module. The attention module just works with the hidden representation and should scale with the length of the sequence? Correct me if i’m wrong but isn’t the computation for each x_t in your hidden sequence, we calculate `a_t = V_a^T tanh(s_t W_s + x_t * W_a) ` and then `h_t = sum(a_t x_t)` The actual increase in the parameters is due to `W_o h_t` When getting the output.

**Comment by andreyzharkov** [comment]
There are ~major changes, so instead of contributing I decided to create another repository instead

**Comment by zafarali** [comment]
Wow this is really neat, I'm going to take a closer look at it. Thank you!
